File: Housing.m
%% Housing Group 8 - Housing Dataset
% Author: Arian Milicija
% Goal: Perform simple cleaning, summaries, and visuals on housing data
% Dataset: Housing.xlsx (sheet 'in')

clear; clc; close all;

%% 1) Load dataset
file = 'Housing.xlsx';
sheet = 'in';
T = readtable(file, 'Sheet', sheet);

%% 2) Handle missing or invalid values
% Check missing counts
disp('Missing values per column:');
disp(sum(ismissing(T)));

% Convert numeric columns to double (if needed)
numVars = {'price','area','bedrooms','bathrooms','stories','parking'};
for i = 1:numel(numVars)
    if iscell(T.(numVars{i})) || isstring(T.(numVars{i})) || ischar(T.(numVars{i}))
        T.(numVars{i}) = str2double(string(T.(numVars{i})));
    end
end

% Drop any rows missing a price or area (key features)
Tclean = T(~(ismissing(T.price) | ismissing(T.area)), :);

fprintf('✅ Cleaned dataset: %d rows remaining.\n', height(Tclean));

%% 3) Summary statistics for Price (like "Total Spent" example)
x = Tclean.price;
fprintf('\n=== Summary Statistics for Price ===\n');
fprintf('Count: %d\n', numel(x));
fprintf('Mean: %.2f\n', mean(x,'omitnan'));
fprintf('Std: %.2f\n', std(x,'omitnan'));
fprintf('Min: %.2f\n', min(x));
fprintf('Median: %.2f\n', median(x));
fprintf('Max: %.2f\n', max(x));
fprintf('Sum: %.2f\n', sum(x));

%% 4) Most common features (analogue to "mostly sold item / payment method")
% Mode for categorical variables
fprintf('\n=== Most Common Features ===\n');
fprintf('Most common furnishing status: %s\n', string(mode(categorical(Tclean.furnishingstatus))));
fprintf('Most common air conditioning: %s\n', string(mode(categorical(Tclean.airconditioning))));
fprintf('Most common basement presence: %s\n', string(mode(categorical(Tclean.basement))));

%% 5) Visualizations

% Histogram of Price
figure('Name','Price Distribution');
histogram(Tclean.price, 20);
xlabel('Price'); ylabel('Count'); title('Distribution of House Prices');
grid on;

% Scatter: Area vs Price
figure('Name','Area vs Price');
scatter(Tclean.area, Tclean.price, 30, 'filled');
xlabel('Area'); ylabel('Price');
title('House Price vs Area');
grid on;

% Boxplot: Price by Furnishing Status
figure('Name','Price by Furnishing');
boxchart(categorical(Tclean.furnishingstatus), Tclean.price);
xlabel('Furnishing Status'); ylabel('Price');
title('Price by Furnishing Status');
grid on;

% Bar Chart: Average Price by Number of Bedrooms
bedStats = groupsummary(Tclean, "bedrooms", "mean", "price");
figure('Name','Average Price by Bedrooms');
bar(bedStats.bedrooms, bedStats.mean_price);
xlabel('Bedrooms'); ylabel('Average Price');
title('Average Price by Number of Bedrooms');
grid on;

% Pie Chart: Furnishing Status
counts = groupcounts(categorical(Tclean.furnishingstatus));
labels = categories(categorical(Tclean.furnishingstatus));
figure('Name','Pie - Furnishing Status');
pie(counts, labels);
title('Furnishing Status Distribution');

% Bar Chart: Parking vs Average Price
parkStats = groupsummary(Tclean, "parking", "mean", "price");
figure('Name','Average Price by Parking Spots');
bar(parkStats.parking, parkStats.mean_price);
xlabel('Parking Spots'); ylabel('Average Price');
title('Average Price by Number of Parking Spots');
grid on;

%% 6) Save outputs
if ~exist('output','dir'), mkdir('output'); end
writetable(Tclean, fullfile('output','Housing_Clean.csv'));
disp('Clean dataset saved to /output/Housing_Clean.csv');

File: Housing.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

# ---------------------------------------------------------
# 1) Load dataset
# ---------------------------------------------------------
file = "Housing.xlsx"
sheet = "in"

T = pd.read_excel(file, sheet_name=sheet)

# ---------------------------------------------------------
# 2) Handle missing or invalid values
# ---------------------------------------------------------
print("Missing values per column:")
print(T.isna().sum())

# Convert numeric columns to float if they came in as strings
numVars = ['price','area','bedrooms','bathrooms','stories','parking']

for col in numVars:
    if T[col].dtype == object:     # string-like
        T[col] = pd.to_numeric(T[col], errors='coerce')

# Drop rows missing key variables
Tclean = T.dropna(subset=['price', 'area'])
print(f"✅ Cleaned dataset: {len(Tclean)} rows remaining.")

# ---------------------------------------------------------
# 3) Summary statistics for Price
# ---------------------------------------------------------
x = Tclean['price']

print("\n=== Summary Statistics for Price ===")
print(f"Count: {x.count()}")
print(f"Mean: {x.mean():.2f}")
print(f"Std: {x.std():.2f}")
print(f"Min: {x.min():.2f}")
print(f"Median: {x.median():.2f}")
print(f"Max: {x.max():.2f}")
print(f"Sum: {x.sum():.2f}")

# ---------------------------------------------------------
# 4) Most common features
# ---------------------------------------------------------
print("\n=== Most Common Features ===")
print("Most common furnishing status:", Tclean["furnishingstatus"].mode()[0])
print("Most common air conditioning:", Tclean["airconditioning"].mode()[0])
print("Most common basement presence:", Tclean["basement"].mode()[0])

# ---------------------------------------------------------
# 5) Visualizations
# ---------------------------------------------------------
plt.figure(figsize=(7,5))
plt.hist(Tclean["price"], bins=20)
plt.xlabel("Price")
plt.ylabel("Count")
plt.title("Distribution of House Prices")
plt.grid(True)
plt.show()

# Scatter: Area vs Price
plt.figure(figsize=(7,5))
plt.scatter(Tclean["area"], Tclean["price"])
plt.xlabel("Area")
plt.ylabel("Price")
plt.title("House Price vs Area")
plt.grid(True)
plt.show()

# Boxplot: Price by Furnishing
plt.figure(figsize=(7,5))
Tclean.boxplot(column="price", by="furnishingstatus")
plt.title("Price by Furnishing Status")
plt.suptitle("")  # remove default pandas title
plt.xlabel("Furnishing Status")
plt.ylabel("Price")
plt.grid(True)
plt.show()

# Bar chart: Average Price by Bedrooms
bedStats = Tclean.groupby("bedrooms")["price"].mean().reset_index()

plt.figure(figsize=(7,5))
plt.bar(bedStats["bedrooms"], bedStats["price"])
plt.xlabel("Bedrooms")
plt.ylabel("Average Price")
plt.title("Average Price by Number of Bedrooms")
plt.grid(True)
plt.show()

# Pie chart: Furnishing Status
furn_counts = Tclean["furnishingstatus"].value_counts()

plt.figure(figsize=(6,6))
plt.pie(furn_counts, labels=furn_counts.index, autopct="%1.1f%%")
plt.title("Furnishing Status Distribution")
plt.show()

# Bar chart: Parking vs Avg Price
parkStats = Tclean.groupby("parking")["price"].mean().reset_index()

plt.figure(figsize=(7,5))
plt.bar(parkStats["parking"], parkStats["price"])
plt.xlabel("Parking Spots")
plt.ylabel("Average Price")
plt.title("Average Price by Number of Parking Spots")
plt.grid(True)
plt.show()

# ---------------------------------------------------------
# 6) Save output
# ---------------------------------------------------------
os.makedirs("output", exist_ok=True)
Tclean.to_csv("output/Housing_Clean.csv", index=False)

print("Clean dataset saved to output/Housing_Clean.csv")
